File: src/code/trie/trie.ts
export function useTrie() {
  class TrieNode {
    children: TrieNode[] = []
    isEnd = false
  }

  const root = new TrieNode()

  function insert(word: string) {
    let node = root
    for (let i = 0; i < word.length; i++) {
      const c = word[i].charCodeAt(0) - 'a'.charCodeAt(0)
      if (!node.children[c])
        node.children[c] = new TrieNode()

      node = node.children[c]
    }
    node.isEnd = true
  }

  function search(word: string): boolean {
    let node = root
    for (let i = 0; i < word.length; i++) {
      const c = word[i].charCodeAt(0) - 'a'.charCodeAt(0)
      if (!node.children[c])
        return false

      node = node.children[c]
    }
    return node.isEnd
  }

  function startsWith(prefix: string): boolean {
    let node = root
    for (let i = 0; i < prefix.length; i++) {
      const c = prefix[i].charCodeAt(0) - 'a'.charCodeAt(0)
      if (!node.children[c])
        return false

      node = node.children[c]
    }
    return true
  }

  return {
    insert,
    search,
    startsWith,
  }
}

File: src/code/trie/trie.rs
pub struct Trie {
    root: TrieNode,
}

impl Trie {
    pub fn new() -> Self {
        Trie {
            root: TrieNode::new(),
        }
    }

    pub fn insert(&mut self, word: String) {
        let mut node = &mut self.root;
        for c in word.chars() {
            let index = (c as u8 - b'a') as usize;
            node = node.children[index].get_or_insert_with(|| Box::new(TrieNode::new()));
        }
        node.is_end = true;
    }

    pub fn search(&self, word: String) -> bool {
        let mut node = &self.root;
        for c in word.chars() {
            let index = (c as u8 - b'a') as usize;
            if let Some(child) = &node.children[index] {
                node = child;
            } else {
                return false;
            }
        }
        node.is_end
    }

    pub fn starts_with(&self, prefix: String) -> bool {
        let mut node = &self.root;
        for c in prefix.chars() {
            let index = (c as u8 - b'a') as usize;
            if let Some(child) = &node.children[index] {
                node = child;
            } else {
                return false;
            }
        }
        true
    }
}

#[derive(Clone)]
struct TrieNode {
    children: Vec<Option<Box<TrieNode>>>,
    is_end: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: vec![None.clone(); 26],
            is_end: false,
        }
    }
}
